File: MoireStudies/MoireView.m
//
//  MoireView.m
//  MoireStudies
//
//  Created by Jialiang Xiang on 2020-12-08.
//

#import "MoireView.h"
#import "PatternView.h"
#import "ControlView.h"
#import "SliderControlView.h"

@implementation MoireView {
    PatternView* patternView1;
    ControlView* controlView1;
    PatternView* patternView2;
    ControlView* controlView2;
    UIView* mask1;
}

- (id) initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        // TODO: stub
        self.backgroundColor = [UIColor blueColor];
    }
    return self;
}

- (void) setUp {
    CGRect controlFrame1 = CGRectMake(10, 30, 200, 300);
    patternView1 = [[PatternView alloc]initWithFrame:CGRectMake(0, 0, self.frame.size.width, self.frame.size.height)];
    patternView1.backgroundColor = [UIColor yellowColor];
    [self addSubview:patternView1];
    // TODO
    [self setUpMaskOnPatternView:patternView2 WithControlFrame:controlFrame1];
    [self setUpControlsWithControlFrame:controlFrame1];
}

- (void)setUpMaskOnPatternView: (UIView*)pv WithControlFrame: (CGRect)c {
    // TODO: mask corresponding pattern view to match the control views
}

- (void)setUpControlsWithControlFrame: (CGRect)c {
    controlView1 = [[SliderControlView alloc]initWithFrame:c];
    [self addSubview:controlView1];
}

@end

File: MoireStudies/MoireView.swift
//
//  MoireView.swift
//  MoireStudies
//
//  Created by Jialiang Xiang on 2020-12-17.
//

import UIKit
import Foundation

class MoireView: UIView {
    typealias PatternViewSubclass = CoreAnimPatternView
    typealias ControlViewSubclass = SliderControlView
    
    var controlFrame1 = CGRect(x: 10, y: 30, width: 200, height: 300)
    var patternView1: PatternView?
    var controlView1: ControlView?
    var patternView2: PatternView?
    var controlView2: ControlView?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.gray
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    @objc func displayMoire(patterns: NSMutableArray) {
        patternView1 = PatternViewSubclass.init(frame: CGRect(x: 0, y: 0, width: self.frame.width, height: self.frame.height))
        if let pv = patternView1 {
            self.addSubview(pv)
            pv.setUp()
        }
        // TODO: setup patternView2
        //self.setUpMaskOnPatternView(patternView: patternView2, controlFrame: controlFrame1)
        self.setUpControls(frame: controlFrame1)
    }
    
    func setUpMaskOnPatternView(patternView: UIView, controlFrame: CGRect) {
        // TODO: mask corresponding pattern view to match the control views
    }
    
    func setUpControls(frame: CGRect) {
        controlView1 = ControlViewSubclass.init(frame: frame)
        if let cv = controlView1 {
            self.addSubview(cv)
        }
    }
}
